src/python/pants/help/help_info_extracter.py

```diff
@@ -249,7 +360,8 @@ class AllHelpInfo:
     scope_to_help_info: dict[str, OptionScopeHelpInfo]
     name_to_goal_info: dict[str, GoalHelpInfo]
     name_to_target_type_info: dict[str, TargetTypeHelpInfo]
-    rule_output_type_to_rule_infos: dict[str, tuple[RuleInfo, ...]]
+    name_to_rule_info: dict[str, RuleInfo]
+    name_to_api_type_info: dict[str, PluginAPITypeInfo]
```

Comment:
Rather than mapping a single output type to a bunch of rules, we instead have two sets of info, one for all rules, and another for all types.